netpyne/batch.py

```diff
@@ -42,38 +58,231 @@ def tupleToStr (obj):
+            jobString = bashTemplate('hpc_slurm') % (jobName, allocation, walltime, nodes, coresPerNode, simDataPath, simDataPath, email, res, custom, simDataFolder, command)
+        # ----------------------------------------------------------------------
+        # run on HPC through PBS
+        # ----------------------------------------------------------------------
+        elif type=='hpc_torque':
+            executer = 'qsub'
+            queueName = args.get('queueName', 'default')
+            nodesppn = 'nodes=%d:ppn=%d' % (nodes, coresPerNode)
+            jobString = bashTemplate('hpc_torque') % (jobName, walltime, queueName, nodesppn, jobName, jobName, custom, command)
+        # ----------------------------------------------------------------------
+        # continue
+        # ----------------------------------------------------------------------
+        print 'Submitting job ', jobName
+        print jobString
+        print '-'*80
+        # save file 
+        batchfile = '%s.sbatch' % (simDataPath)
+        with open(batchfile, 'w') as text_file:
+            text_file.write("%s" % jobString)
+        
+        with open(simDataPath+'.run', 'w') as outf, open(simDataPath+'.err', 'w') as errf:
+            pids.append(Popen([executer, batchfile], stdout=outf,  stderr=errf, preexec_fn=os.setsid).pid)
+        total_jobs += 1
+        sleep(0.1)
+    # ----------------------------------------------------------------------
+    # gather data and compute fitness
+    # ----------------------------------------------------------------------
+    num_iters = 0
+    jobs_completed = 0
+    targetFitness = [None for cand in candidates]
+    # print outfilestem
+    print "waiting jobs from generation: %d/%d to finish..." %(ngen, args.get('max_generations'))
+    print "PID's: %r" %(pids)
+    # start fitness calculation
+    while jobs_completed < total_jobs:
+        unfinished = [i for i, x in enumerate(targetFitness) if x is None ]
+        for candidate_index in unfinished:
+            try: # load simData and evaluate fitness
+                simDataPath = simDataFolder + "/gen_" + str(ngen) + "_cand_" + str(candidate_index)
+                with open('%s.json'% (simDataPath)) as file:
+                    simData = json.load(file)['simData']
+                targetFitness[candidate_index] = eval(fitness_expression)
+                jobs_completed += 1
+            except:
+                pass
+        num_iters += 1
+        if num_iters >= args.get('maxiter_wait', 5000): 
+            print "max iterations reached -- remaining jobs set to default fitness"
+            for canditade_index in unfinished:
+                targetFitness[canditade_index] = default_fitness
+                jobs_completed += 1
+                
+        sleep(args.get('time_sleep', 1))
+    # kill all processes
+    for pid in pids: os.killpg(os.getpgid(pid), signal.SIGTERM)
```

Comment:
@salvadord the nrniv jobs remain open after job finished. So I am killing them with this